File: client/csharp/Angzarr.Client/Router.cs
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using Type = System.Type;

namespace Angzarr.Client;

/// <summary>
/// Error message constants.
/// </summary>
public static class ErrorMessages
{
    public const string UnknownCommand = "Unknown command type";
    public const string NoCommandPages = "No command pages";
}

/// <summary>
/// Delegate for command handlers in the functional pattern.
/// </summary>
public delegate Angzarr.EventBook CommandHandler<TState>(
    Angzarr.CommandBook commandBook,
    Any commandAny,
    TState state,
    int seq);

/// <summary>
/// Delegate for state rebuild functions.
/// </summary>
public delegate TState StateRebuilder<TState>(Angzarr.EventBook? eventBook);

/// <summary>
/// Response from rejection handlers - can emit events AND/OR notification.
/// </summary>
public class RejectionHandlerResponse
{
    /// <summary>
    /// Events to persist to own state (compensation).
    /// </summary>
    public Angzarr.EventBook? Events { get; set; }

    /// <summary>
    /// Notification to forward upstream (rejection propagation).
    /// </summary>
    public Angzarr.Notification? Notification { get; set; }
}

/// <summary>
/// Delegate for rejection handlers.
/// </summary>
public delegate RejectionHandlerResponse RejectionHandler<TState>(
    Angzarr.Notification notification,
    TState state);

/// <summary>
/// DRY command dispatcher for aggregates.
/// Matches command type_url suffixes and dispatches to registered handlers.
/// </summary>
public class CommandRouter<TState> where TState : new()
{
    private readonly string _domain;
    private StateRebuilder<TState>? _rebuild;
    private StateRouter<TState>? _stateRouter;
    private readonly List<(string Suffix, CommandHandler<TState> Handler)> _handlers = new();
    private readonly Dictionary<string, RejectionHandler<TState>> _rejectionHandlers = new();

    public CommandRouter(string domain, StateRebuilder<TState>? rebuild = null)
    {
        _domain = domain;
        _rebuild = rebuild;
    }

    /// <summary>
    /// Compose a StateRouter for state reconstruction.
    /// </summary>
    public CommandRouter<TState> WithState(StateRouter<TState> stateRouter)
    {
        _stateRouter = stateRouter;
        return this;
    }

    /// <summary>
    /// Register a handler for a command type_url suffix.
    /// </summary>
    public CommandRouter<TState> On(string suffix, CommandHandler<TState> handler)
    {
        _handlers.Add((suffix, handler));
        return this;
    }

    /// <summary>
    /// Register a rejection handler for compensation.
    /// </summary>
    public CommandRouter<TState> OnRejected(string domain, string command, RejectionHandler<TState> handler)
    {
        _rejectionHandlers[$"{domain}/{command}"] = handler;
        return this;
    }

    /// <summary>
    /// Dispatch a ContextualCommand to the matching handler.
    /// </summary>
    public Angzarr.BusinessResponse Dispatch(Angzarr.ContextualCommand cmd)
    {
        var commandBook = cmd.Command;
        var priorEvents = cmd.Events;

        var state = GetState(priorEvents);
        var seq = Helpers.NextSequence(priorEvents);

        if (commandBook.Pages.Count == 0)
            throw new InvalidArgumentError(ErrorMessages.NoCommandPages);

        var commandAny = commandBook.Pages[0].Command;
        if (string.IsNullOrEmpty(commandAny?.TypeUrl))
            throw new InvalidArgumentError(ErrorMessages.NoCommandPages);

        var typeUrl = commandAny.TypeUrl;

        // Check for Notification (rejection/compensation)
        if (typeUrl.EndsWith("Notification"))
        {
            var notification = commandAny.Unpack<Angzarr.Notification>();
            return DispatchRejection(notification, state);
        }

        // Normal command dispatch
        foreach (var (suffix, handler) in _handlers)
        {
            if (typeUrl.EndsWith(suffix))
            {
                var events = handler(commandBook, commandAny, state, seq);
                return new Angzarr.BusinessResponse { Events = events };
            }
        }

        throw new InvalidArgumentError($"{ErrorMessages.UnknownCommand}: {typeUrl}");
    }

    private Angzarr.BusinessResponse DispatchRejection(Angzarr.Notification notification, TState state)
    {
        var rejection = notification.Payload?.Unpack<Angzarr.RejectionNotification>();
        var domain = "";
        var commandSuffix = "";

        if (rejection?.RejectedCommand?.Pages.Count > 0)
        {
            var rejectedCmd = rejection.RejectedCommand;
            domain = rejectedCmd.Cover?.Domain ?? "";
            var cmdTypeUrl = rejectedCmd.Pages[0].Command?.TypeUrl ?? "";
            commandSuffix = Helpers.TypeNameFromUrl(cmdTypeUrl);
        }

        foreach (var (key, handler) in _rejectionHandlers)
        {
            var parts = key.Split('/');
            if (parts[0] == domain && commandSuffix.EndsWith(parts[1]))
            {
                var response = handler(notification, state);
                // Handle notification forwarding
                if (response.Notification != null)
                {
                    return new Angzarr.BusinessResponse { Notification = response.Notification };
                }
                // Handle compensation events
                if (response.Events != null)
                {
                    return new Angzarr.BusinessResponse { Events = response.Events };
                }
                // Handler returned empty response
                return new Angzarr.BusinessResponse
                {
                    Revocation = new Angzarr.RevocationResponse
                    {
                        EmitSystemRevocation = false,
                        Reason = $"Aggregate {_domain} handled rejection for {key}"
                    }
                };
            }
        }

        return new Angzarr.BusinessResponse
        {
            Revocation = new Angzarr.RevocationResponse
            {
                EmitSystemRevocation = true,
                Reason = $"Aggregate {_domain} has no custom compensation for {domain}/{commandSuffix}"
            }
        };
    }

    private TState GetState(Angzarr.EventBook? eventBook)
    {
        if (_stateRouter != null)
            return _stateRouter.WithEventBook(eventBook);
        if (_rebuild != null)
            return _rebuild(eventBook);
        throw new InvalidOperationException(
            "CommandRouter requires either rebuild function or StateRouter via WithState()");
    }
}

/// <summary>
/// Delegate for event handlers in the functional pattern.
/// </summary>
public delegate List<Angzarr.CommandBook> EventHandler(
    Any eventAny,
    byte[]? root,
    string correlationId,
    List<Angzarr.EventBook>? destinations);

/// <summary>
/// Delegate for prepare handlers in two-phase protocol.
/// </summary>
public delegate List<Angzarr.Cover> PrepareHandler(Any eventAny, Angzarr.UUID? root);

/// <summary>
/// Unified event dispatcher for sagas, process managers, and projectors.
/// Uses fluent .Domain().On() pattern to register handlers with domain context.
///
/// Example (Saga - single domain):
/// <code>
/// var router = new EventRouter("saga-table-hand")
///     .Domain("table")
///     .On("HandStarted", HandleStarted);
/// </code>
///
/// Example (Process Manager - multi-domain):
/// <code>
/// var router = new EventRouter("pmg-order-flow")
///     .Domain("order")
///     .On("OrderCreated", HandleCreated)
///     .Domain("inventory")
///     .On("StockReserved", HandleReserved);
/// </code>
///
/// Example (Projector - multi-domain):
/// <code>
/// var router = new EventRouter("prj-output")
///     .Domain("player")
///     .On("PlayerRegistered", HandleRegistered)
///     .Domain("hand")
///     .On("CardsDealt", HandleDealt);
/// </code>
/// </summary>
public class EventRouter
{
    private readonly string _name;
    private string? _currentDomain;
    private readonly Dictionary<string, List<(string Suffix, EventHandler Handler)>> _handlers = new();
    private readonly Dictionary<string, Dictionary<string, PrepareHandler>> _prepareHandlers = new();

    public EventRouter(string name)
    {
        _name = name;
    }

    /// <summary>
    /// Create a new EventRouter with a single input domain (backwards compatibility).
    /// </summary>
    /// <param name="name">Component name</param>
    /// <param name="inputDomain">Single input domain (deprecated, use Domain() instead)</param>
    [Obsolete("Use new EventRouter(name).Domain(inputDomain) instead")]
    public EventRouter(string name, string inputDomain) : this(name)
    {
        if (!string.IsNullOrEmpty(inputDomain))
            Domain(inputDomain);
    }

    /// <summary>
    /// Set the current domain context for subsequent On() calls.
    /// </summary>
    public EventRouter Domain(string name)
    {
        _currentDomain = name;
        if (!_handlers.ContainsKey(name))
            _handlers[name] = new List<(string, EventHandler)>();
        if (!_prepareHandlers.ContainsKey(name))
            _prepareHandlers[name] = new Dictionary<string, PrepareHandler>();
        return this;
    }

    /// <summary>
    /// Register a prepare handler for an event type_url suffix.
    /// Must be called after Domain() to set context.
    /// </summary>
    public EventRouter Prepare(string suffix, PrepareHandler handler)
    {
        if (_currentDomain == null)
            throw new InvalidOperationException("Must call Domain() before Prepare()");
        _prepareHandlers[_currentDomain][suffix] = handler;
        return this;
    }

    /// <summary>
    /// Register a handler for an event type_url suffix in current domain.
    /// Must be called after Domain() to set context.
    /// </summary>
    public EventRouter On(string suffix, EventHandler handler)
    {
        if (_currentDomain == null)
            throw new InvalidOperationException("Must call Domain() before On()");
        _handlers[_currentDomain].Add((suffix, handler));
        return this;
    }

    /// <summary>
    /// Auto-derive subscriptions from registered handlers.
    /// </summary>
    /// <returns>Dictionary of domain to event types.</returns>
    public Dictionary<string, List<string>> Subscriptions()
    {
        var result = new Dictionary<string, List<string>>();
        foreach (var (domain, handlers) in _handlers)
        {
            if (handlers.Count > 0)
                result[domain] = handlers.Select(h => h.Suffix).ToList();
        }
        return result;
    }

    /// <summary>
    /// Get destinations needed for the given source events.
    /// Routes based on source domain.
    /// </summary>
    public List<Angzarr.Cover> PrepareDestinations(Angzarr.EventBook book)
    {
        var sourceDomain = book.Cover?.Domain ?? "";
        if (!_prepareHandlers.TryGetValue(sourceDomain, out var domainHandlers))
            return new List<Angzarr.Cover>();

        var root = book.Cover?.Root;
        var destinations = new List<Angzarr.Cover>();

        foreach (var page in book.Pages)
        {
            if (page.Event == null) continue;
            foreach (var (suffix, handler) in domainHandlers)
            {
                if (page.Event.TypeUrl.EndsWith(suffix))
                {
                    destinations.AddRange(handler(page.Event, root));
                    break;
                }
            }
        }
        return destinations;
    }

    /// <summary>
    /// Dispatch all events in an EventBook to registered handlers.
    /// Routes based on source domain and event type suffix.
    /// </summary>
    public List<Angzarr.CommandBook> Dispatch(
        Angzarr.EventBook book,
        List<Angzarr.EventBook>? destinations = null)
    {
        var sourceDomain = book.Cover?.Domain ?? "";
        if (!_handlers.TryGetValue(sourceDomain, out var domainHandlers))
            return new List<Angzarr.CommandBook>();

        var root = book.Cover?.Root?.Value.ToByteArray();
        var correlationId = book.Cover?.CorrelationId ?? "";
        var commands = new List<Angzarr.CommandBook>();

        foreach (var page in book.Pages)
        {
            if (page.Event == null) continue;
            foreach (var (suffix, handler) in domainHandlers)
            {
                if (page.Event.TypeUrl.EndsWith(suffix))
                {
                    commands.AddRange(handler(page.Event, root, correlationId, destinations));
                    break;
                }
            }
        }
        return commands;
    }

    /// <summary>
    /// Return the first registered domain (for backwards compatibility).
    /// </summary>
    [Obsolete("Use Subscriptions() instead")]
    public string InputDomain() => _handlers.Keys.FirstOrDefault() ?? "";

}

/// <summary>
/// Delegate for state appliers.
/// </summary>
public delegate void StateApplier<TState>(TState state, IMessage eventMessage);

/// <summary>
/// Fluent state reconstruction from events.
/// </summary>
public class StateRouter<TState> where TState : new()
{
    private readonly Dictionary<string, (Type EventType, Action<TState, Any> Applier)> _appliers = new();

    /// <summary>
    /// Register an event applier.
    /// </summary>
    public StateRouter<TState> On<TEvent>(Action<TState, TEvent> applier) where TEvent : IMessage, new()
    {
        var suffix = typeof(TEvent).Name;
        _appliers[suffix] = (typeof(TEvent), (state, any) =>
        {
            var evt = any.Unpack<TEvent>();
            applier(state, evt);
        });
        return this;
    }

    /// <summary>
    /// Rebuild state from an EventBook.
    /// </summary>
    public TState WithEventBook(Angzarr.EventBook? book)
    {
        var state = new TState();
        if (book == null) return state;

        foreach (var page in book.Pages)
        {
            if (page.Event == null) continue;
            ApplyEvent(state, page.Event);
        }
        return state;
    }

    private void ApplyEvent(TState state, Any eventAny)
    {
        foreach (var (suffix, (_, applier)) in _appliers)
        {
            if (eventAny.TypeUrl.EndsWith(suffix))
            {
                applier(state, eventAny);
                return;
            }
        }
        // Unknown event type - silently ignore (forward compatibility)
    }
}

/// <summary>
/// Delegate for upcaster handlers.
/// </summary>
public delegate Any UpcasterHandler(Any eventAny);

/// <summary>
/// Event version transformer.
/// </summary>
public class UpcasterRouter
{
    private readonly string _domain;
    private readonly List<(string Suffix, UpcasterHandler Handler, Type ToType)> _handlers = new();

    public UpcasterRouter(string domain)
    {
        _domain = domain;
    }

    /// <summary>
    /// Register an upcaster handler.
    /// </summary>
    public UpcasterRouter On(string suffix, UpcasterHandler handler, Type? toType = null)
    {
        _handlers.Add((suffix, handler, toType ?? typeof(object)));
        return this;
    }

    /// <summary>
    /// Transform a list of events to current versions.
    /// </summary>
    public List<Angzarr.EventPage> Upcast(IEnumerable<Angzarr.EventPage> events)
    {
        var result = new List<Angzarr.EventPage>();

        foreach (var page in events)
        {
            if (page.Event == null)
            {
                result.Add(page);
                continue;
            }

            var transformed = false;
            foreach (var (suffix, handler, _) in _handlers)
            {
                if (page.Event.TypeUrl.EndsWith(suffix))
                {
                    var newEvent = handler(page.Event);
                    var newPage = new Angzarr.EventPage
                    {
                        Event = newEvent,
                        Sequence = page.Sequence,
                        CreatedAt = page.CreatedAt
                    };
                    result.Add(newPage);
                    transformed = true;
                    break;
                }
            }

            if (!transformed)
                result.Add(page);
        }

        return result;
    }
}

File: client/java/client/src/main/java/dev/angzarr/client/Router.java
package dev.angzarr.client;

import com.google.protobuf.Any;
import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.Message;
import dev.angzarr.*;

import dev.angzarr.client.compensation.RejectionHandlerResponse;

import java.util.*;
import java.util.function.BiFunction;
import java.util.function.Function;

/**
 * DRY command dispatcher for aggregates (functional pattern).
 */
class CommandRouter<S> {
    private final String domain;
    private Function<EventBook, S> rebuild;
    private StateRouter<S> stateRouter;
    private final List<Map.Entry<String, CommandHandler<S>>> handlers = new ArrayList<>();
    private final Map<String, RejectionHandler<S>> rejectionHandlers = new HashMap<>();

    @FunctionalInterface
    public interface CommandHandler<S> {
        EventBook handle(CommandBook commandBook, Any commandAny, S state, int seq);
    }

    @FunctionalInterface
    public interface RejectionHandler<S> {
        RejectionHandlerResponse handle(Notification notification, S state);
    }

    public CommandRouter(String domain) {
        this.domain = domain;
    }

    public CommandRouter(String domain, Function<EventBook, S> rebuild) {
        this.domain = domain;
        this.rebuild = rebuild;
    }

    public CommandRouter<S> withState(StateRouter<S> stateRouter) {
        this.stateRouter = stateRouter;
        return this;
    }

    public CommandRouter<S> on(String suffix, CommandHandler<S> handler) {
        handlers.add(new AbstractMap.SimpleEntry<>(suffix, handler));
        return this;
    }

    public CommandRouter<S> onRejected(String domain, String command, RejectionHandler<S> handler) {
        rejectionHandlers.put(domain + "/" + command, handler);
        return this;
    }

    public BusinessResponse dispatch(ContextualCommand cmd) {
        var commandBook = cmd.getCommand();
        var priorEvents = cmd.hasEvents() ? cmd.getEvents() : null;

        var state = getState(priorEvents);
        var seq = Helpers.nextSequence(priorEvents);

        if (commandBook.getPagesList().isEmpty()) {
            throw new Errors.InvalidArgumentError("No command pages");
        }

        var commandAny = commandBook.getPages(0).getCommand();
        if (commandAny.getTypeUrl().isEmpty()) {
            throw new Errors.InvalidArgumentError("No command pages");
        }

        var typeUrl = commandAny.getTypeUrl();

        // Check for Notification
        if (typeUrl.endsWith("Notification")) {
            try {
                var notification = commandAny.unpack(Notification.class);
                return dispatchRejection(notification, state);
            } catch (InvalidProtocolBufferException e) {
                throw new Errors.ClientError("Failed to unpack notification", e);
            }
        }

        // Normal command dispatch
        for (var entry : handlers) {
            if (typeUrl.endsWith(entry.getKey())) {
                var events = entry.getValue().handle(commandBook, commandAny, state, seq);
                return BusinessResponse.newBuilder().setEvents(events).build();
            }
        }

        throw new Errors.InvalidArgumentError("Unknown command type: " + typeUrl);
    }

    private BusinessResponse dispatchRejection(Notification notification, S state) {
        String domain = "";
        String commandSuffix = "";

        if (notification.hasPayload()) {
            try {
                var rejection = notification.getPayload().unpack(RejectionNotification.class);
                if (rejection.hasRejectedCommand() && !rejection.getRejectedCommand().getPagesList().isEmpty()) {
                    var rejectedCmd = rejection.getRejectedCommand();
                    domain = rejectedCmd.hasCover() ? rejectedCmd.getCover().getDomain() : "";
                    var cmdTypeUrl = rejectedCmd.getPages(0).getCommand().getTypeUrl();
                    commandSuffix = Helpers.typeNameFromUrl(cmdTypeUrl);
                }
            } catch (InvalidProtocolBufferException ignored) {}
        }

        for (var entry : rejectionHandlers.entrySet()) {
            var parts = entry.getKey().split("/");
            if (parts[0].equals(domain) && commandSuffix.endsWith(parts[1])) {
                var response = entry.getValue().handle(notification, state);
                // Handle notification forwarding
                if (response.hasNotification()) {
                    return BusinessResponse.newBuilder()
                        .setNotification(response.getNotification())
                        .build();
                }
                // Handle compensation events
                if (response.hasEvents()) {
                    return BusinessResponse.newBuilder()
                        .setEvents(response.getEvents())
                        .build();
                }
                // Handler returned empty response
                return BusinessResponse.newBuilder()
                    .setRevocation(RevocationResponse.newBuilder()
                        .setEmitSystemRevocation(false)
                        .setReason("Aggregate " + this.domain + " handled rejection for " + entry.getKey())
                        .build())
                    .build();
            }
        }

        return BusinessResponse.newBuilder()
            .setRevocation(RevocationResponse.newBuilder()
                .setEmitSystemRevocation(true)
                .setReason("Aggregate " + this.domain + " has no custom compensation for " + domain + "/" + commandSuffix)
                .build())
            .build();
    }

    private S getState(EventBook eventBook) {
        if (stateRouter != null) {
            return stateRouter.withEventBook(eventBook);
        }
        if (rebuild != null) {
            return rebuild.apply(eventBook);
        }
        throw new IllegalStateException("CommandRouter requires either rebuild function or StateRouter");
    }
}

/**
 * Unified event dispatcher for sagas, process managers, and projectors.
 * Uses fluent .domain().on() pattern to register handlers with domain context.
 *
 * Example (Saga - single domain):
 * <pre>
 * EventRouter router = new EventRouter("saga-table-hand")
 *     .domain("table")
 *     .on("HandStarted", this::handleStarted);
 * </pre>
 *
 * Example (Process Manager - multi-domain):
 * <pre>
 * EventRouter router = new EventRouter("pmg-order-flow")
 *     .domain("order")
 *     .on("OrderCreated", this::handleCreated)
 *     .domain("inventory")
 *     .on("StockReserved", this::handleReserved);
 * </pre>
 *
 * Example (Projector - multi-domain):
 * <pre>
 * EventRouter router = new EventRouter("prj-output")
 *     .domain("player")
 *     .on("PlayerRegistered", this::handleRegistered)
 *     .domain("hand")
 *     .on("CardsDealt", this::handleDealt);
 * </pre>
 */
class EventRouter {
    private final String name;
    private String currentDomain;
    private final Map<String, List<Map.Entry<String, EventHandler>>> handlers = new HashMap<>();
    private final Map<String, Map<String, PrepareHandler>> prepareHandlers = new HashMap<>();

    @FunctionalInterface
    public interface EventHandler {
        List<CommandBook> handle(Any eventAny, byte[] root, String correlationId, List<EventBook> destinations);
    }

    @FunctionalInterface
    public interface PrepareHandler {
        List<Cover> handle(Any eventAny, dev.angzarr.UUID root);
    }

    public EventRouter(String name) {
        this.name = name;
    }

    /**
     * Create a new EventRouter with a single input domain (backwards compatibility).
     * @deprecated Use new EventRouter(name).domain(inputDomain) instead.
     */
    @Deprecated
    public EventRouter(String name, String inputDomain) {
        this.name = name;
        if (inputDomain != null && !inputDomain.isEmpty()) {
            domain(inputDomain);
        }
    }

    /**
     * Set the current domain context for subsequent on() calls.
     */
    public EventRouter domain(String name) {
        this.currentDomain = name;
        handlers.computeIfAbsent(name, k -> new ArrayList<>());
        prepareHandlers.computeIfAbsent(name, k -> new HashMap<>());
        return this;
    }

    /**
     * Register a prepare handler for an event type_url suffix.
     * Must be called after domain() to set context.
     */
    public EventRouter prepare(String suffix, PrepareHandler handler) {
        if (currentDomain == null) {
            throw new IllegalStateException("Must call domain() before prepare()");
        }
        prepareHandlers.get(currentDomain).put(suffix, handler);
        return this;
    }

    /**
     * Register a handler for an event type_url suffix in current domain.
     * Must be called after domain() to set context.
     */
    public EventRouter on(String suffix, EventHandler handler) {
        if (currentDomain == null) {
            throw new IllegalStateException("Must call domain() before on()");
        }
        handlers.get(currentDomain).add(new AbstractMap.SimpleEntry<>(suffix, handler));
        return this;
    }

    /**
     * Auto-derive subscriptions from registered handlers.
     * @return Map of domain to event types.
     */
    public Map<String, List<String>> subscriptions() {
        var result = new HashMap<String, List<String>>();
        for (var entry : handlers.entrySet()) {
            if (!entry.getValue().isEmpty()) {
                result.put(entry.getKey(),
                    entry.getValue().stream().map(Map.Entry::getKey).toList());
            }
        }
        return result;
    }

    /**
     * Get destinations needed for the given source events.
     * Routes based on source domain.
     */
    public List<Cover> prepareDestinations(EventBook book) {
        var sourceDomain = book.hasCover() ? book.getCover().getDomain() : "";
        var domainHandlers = prepareHandlers.get(sourceDomain);
        if (domainHandlers == null) {
            return List.of();
        }

        var root = book.hasCover() ? book.getCover().getRoot() : null;
        var destinations = new ArrayList<Cover>();

        for (var page : book.getPagesList()) {
            if (!page.hasEvent()) continue;
            for (var entry : domainHandlers.entrySet()) {
                if (page.getEvent().getTypeUrl().endsWith(entry.getKey())) {
                    destinations.addAll(entry.getValue().handle(page.getEvent(), root));
                    break;
                }
            }
        }
        return destinations;
    }

    /**
     * Dispatch all events in an EventBook to registered handlers.
     * Routes based on source domain and event type suffix.
     */
    public List<CommandBook> dispatch(EventBook book, List<EventBook> destinations) {
        var sourceDomain = book.hasCover() ? book.getCover().getDomain() : "";
        var domainHandlers = handlers.get(sourceDomain);
        if (domainHandlers == null) {
            return List.of();
        }

        var root = book.hasCover() ? book.getCover().getRoot().getValue().toByteArray() : null;
        var correlationId = book.hasCover() ? book.getCover().getCorrelationId() : "";
        var dests = destinations != null ? destinations : List.<EventBook>of();

        var commands = new ArrayList<CommandBook>();
        for (var page : book.getPagesList()) {
            if (!page.hasEvent()) continue;
            for (var entry : domainHandlers) {
                if (page.getEvent().getTypeUrl().endsWith(entry.getKey())) {
                    commands.addAll(entry.getValue().handle(page.getEvent(), root, correlationId, dests));
                    break;
                }
            }
        }
        return commands;
    }

    /**
     * Return the first registered domain (for backwards compatibility).
     * @deprecated Use subscriptions() instead.
     */
    @Deprecated
    public String inputDomain() {
        return handlers.keySet().stream().findFirst().orElse("");
    }

    /**
     * Declare an output domain and command type (deprecated, no-op).
     * This method was used for topology discovery but is now deprecated.
     * @deprecated This method has no effect and will be removed.
     */
    @Deprecated
    public EventRouter sends(String domain, String commandType) {
        // No-op for backwards compatibility
        return this;
    }

    /**
     * Return output domain names (deprecated, returns empty list).
     * @deprecated Output domains are no longer tracked.
     */
    @Deprecated
    public List<String> outputDomains() {
        return List.of();
    }

    /**
     * Return command types for a given output domain (deprecated, returns empty list).
     * @deprecated Output types are no longer tracked.
     */
    @Deprecated
    public List<String> outputTypes(String domain) {
        return List.of();
    }
}

/**
 * Fluent state reconstruction from events (functional pattern).
 */
class StateRouter<S> {
    private final java.util.function.Supplier<S> factory;
    private final Map<String, BiFunction<S, Any, Void>> appliers = new HashMap<>();

    public StateRouter(java.util.function.Supplier<S> factory) {
        this.factory = factory;
    }

    public <E extends Message> StateRouter<S> on(Class<E> eventType, java.util.function.BiConsumer<S, E> applier) {
        var suffix = eventType.getSimpleName();
        appliers.put(suffix, (state, any) -> {
            try {
                @SuppressWarnings("unchecked")
                E event = (E) any.unpack(eventType);
                applier.accept(state, event);
            } catch (InvalidProtocolBufferException ignored) {}
            return null;
        });
        return this;
    }

    public S withEventBook(EventBook book) {
        var state = factory.get();
        if (book == null) return state;

        for (var page : book.getPagesList()) {
            if (!page.hasEvent()) continue;
            applyEvent(state, page.getEvent());
        }
        return state;
    }

    private void applyEvent(S state, Any eventAny) {
        for (var entry : appliers.entrySet()) {
            if (eventAny.getTypeUrl().endsWith(entry.getKey())) {
                entry.getValue().apply(state, eventAny);
                return;
            }
        }
    }
}
